protocol/contracts/lpp/src/state/total.rs

```diff
@@ -81,14 +81,24 @@ where
     ) -> Result<&Self, ContractError> {
         self.total_interest_due = self.total_interest_due_by_now(&ctime);
 
+        let updated_total_principal_due =
+            self.total_principal_due
+                .checked_add(amount)
+                .ok_or(ContractError::OverflowError(
+                    "Total principal due overflow".to_string(),
+                ))?;
+
         // TODO: get rid of fully qualified syntax
         self.annual_interest_rate = Rational::new(
             Fraction::<Coin<Lpn>>::of(&self.annual_interest_rate, self.total_principal_due)
-                + loan_interest_rate.of(amount),
-            self.total_principal_due + amount,
+                .checked_add(loan_interest_rate.of(amount))
+                .ok_or(ContractError::OverflowError(
+                    "Annual interest rate calculation overflow".to_string(),
+                ))?,
```

Comment:
the code may become clearer if you extract the nominator as a separate variable - and name it appropriately - that is the most important! The goal is to end up with code that is easy to read and comprehend